flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala

```diff
@@ -944,7 +956,12 @@ class Table(
     * @param tableName Name of the registered [[TableSink]] to which the [[Table]] is written.
     */
   def insertInto(tableName: String): Unit = {
-    tableEnv.insertInto(this, tableName, this.tableEnv.queryConfig)
+    this.logicalPlan match {
+      case _: LogicalTableFunctionCall =>
+        throw new ValidationException("TableFunction can only be used in join and leftOuterJoin.")
+      case _ =>
+        tableEnv.insertInto(this, tableName, this.tableEnv.queryConfig)
```

Comment:
Can we using "insertInto(tableName, this.tableEnv.queryConfig)", that share the validation logic ?

Reply:
This doesn't work because `this.tableEnv` is null if the Table was generated from a `TableFunctionCall`. 
So we cannot call `insertInto(tableName, this.tableEnv.queryConfig)` because it throws an NPE.

Reply:
I feel like that put same code in the two `overload` method is not the best way.  We should check the NULL of `tableEnv` Before we call `insertInto(tableName, this.tableEnv.queryConfig)`, It's that make sense to you?

Reply:
The check is basically replacing the null check but provides more context. 
What would we do if `tableEnv` is `null`? Do nothing? Throw an exception? What  would be the error message? By checking for the type of the root node of the logical plan we know why the call would not succeed and can give an appropriate error message. 

We could of course also pass `null` as configuration and check for it to be null in `insertInto(String, QueryConfig)` but I'd consider that as an invalid use of the interface that should also result in an exception.

Reply:
if we check NULL, we should throw exception, but I think your right. I agree keep the current approach .